src/staticdata_utils.c

```diff
@@ -1001,61 +1024,59 @@ static void jl_insert_backedges(jl_array_t *edges, jl_array_t *ext_ci_list, size
-                        jl_method_table_add_backedge(mt, sig, codeinst);
-                        j += 2;
-                    }
-                    else {
-                        jl_value_t *callee = jl_svecref(callees, j + 1);
-                        if (jl_is_code_instance(callee))
-                            callee = (jl_value_t*)((jl_code_instance_t*)callee)->def;
-                        else if (jl_is_method(callee)) {
-                            j += 2;
+            jl_verify_method_graph(codeinst, &stack, &visiting);
+            size_t minvalid = jl_atomic_load_relaxed(&codeinst->min_world);
+            size_t maxvalid = jl_atomic_load_relaxed(&codeinst->max_world);
+            if (maxvalid >= minvalid) {
+                if (jl_atomic_load_relaxed(&jl_world_counter) == maxvalid) {
+                    // if this callee is still valid, add all the backedges
+                    for (size_t j = 0; j < jl_svec_len(callees); ) {
+                        jl_value_t *edge = jl_svecref(callees, j);
+                        if (jl_is_long(edge)) {
+                            j += 2; // skip over signature and count but not methods
                             continue;
                         }
-                        jl_method_instance_add_backedge((jl_method_instance_t*)callee, edge, codeinst);
-                        j += 2;
+                        else if (jl_is_method(edge)) {
+                            j += 1;
+                            continue;
+                        }
+                        if (jl_is_code_instance(edge))
+                            edge = (jl_value_t*)((jl_code_instance_t*)edge)->def;
+                        if (jl_is_method_instance(edge)) {
+                            jl_method_instance_add_backedge((jl_method_instance_t*)edge, NULL, codeinst);
+                            j += 1;
+                        }
+                        else if (jl_is_mtable(edge)) {
+                            jl_methtable_t *mt = (jl_methtable_t*)edge;
+                            jl_value_t *sig = jl_svecref(callees, j + 1);
+                            jl_method_table_add_backedge(mt, sig, codeinst);
+                            j += 2;
+                        }
+                        else {
+                            jl_value_t *callee = jl_svecref(callees, j + 1);
+                            if (jl_is_code_instance(callee))
+                                callee = (jl_value_t*)((jl_code_instance_t*)callee)->def;
+                            else if (jl_is_method(callee)) {
+                                j += 2;
+                                continue;
+                            }
+                            jl_method_instance_add_backedge((jl_method_instance_t*)callee, edge, codeinst);
+                            j += 2;
+                        }
                     }
                 }
+                if (jl_atomic_load_relaxed(&jl_world_counter) == maxvalid) {
```

Comment:
This check seems to always hold true as the equivalent check is done just above?

Reply:
Time has passed between them, so what was once true may be no longer